topology/probes/bess/bess.go

```diff
@@ -0,0 +1,248 @@
+			}
+		}
+		if !topology.HaveOwnershipLink(p.Ctx.Graph, p.Ctx.RootNode, bessNode) {
+			topology.AddOwnershipLink(p.Ctx.Graph, p.Ctx.RootNode, bessNode, nil)
+		}
+
+		// find ports to figure out pipeline roots
+		portMap := make(map[string]*graph.Node)
+		var bessPort *graph.Node
+		for _, port := range portRes.GetPorts() {
+			portID := graph.GenID(port.GetName())
+			bessPort = p.Ctx.Graph.GetNode(portID)
+			if bessPort == nil {
+				bessPort, err = p.Ctx.Graph.NewNode(portID, graph.Metadata{
+					"Name":    port.GetName(),
+					"MAC":     port.GetMacAddr(),
+					"Type":    port.GetDriver(),
+					"Backend": BessBackend,
+				})
+				if err != nil {
+					log.Errorf("Failed to create bessPort: %v", bessPort)
+				}
+			}
+			if !topology.HaveOwnershipLink(p.Ctx.Graph, bessNode, bessPort) {
+				topology.AddOwnershipLink(p.Ctx.Graph, bessNode, bessPort, nil)
+			}
+			portMap[port.GetName()] = bessPort
+		}
+
+		// find modules and associate to pipelines/ports
+		for _, module := range modsRes.GetModules() {
+			if module.GetName() == "" {
+				continue
+			}
+			moduleName := module.GetName()
+			// find first modules in pipeline
+			if module.GetMclass() == "PortInc" {
+				p.walkAndRender(moduleName, portMap[strings.TrimSuffix(moduleName, "_ingress")], &foundEdges)
+			}
+			foundMods = append(foundMods, moduleName)
+		}
+
+		// if there are any remaining modules then they do not belong to a port pipeline, render them anyway,
+		// log warning, parent node will be bess node itself
+		for _, module := range foundMods {
+			node := p.Ctx.Graph.GetNode(graph.GenID(module))
+			if node == nil {
+				log.Warningf("Node found in bess that is not part of any port pipeline: %s", module)
+				p.walkAndRender(module, bessNode, &foundEdges)
+			}
+		}
+
+		// Now check for any nodes that need to be removed
+		filter := graph.NewElementFilter(filters.NewTermStringFilter("Backend", BessBackend))
+		nodes := p.Ctx.Graph.GetNodes(filter)
+		// foundNodes will contain all nodes we know should exist
+		var foundNodes []string
+		foundNodes = append(foundNodes, foundMods...)
+		for k := range portMap {
+			foundNodes = append(foundNodes, k)
+		}
+
+		for _, node := range nodes {
+			modFound := false
+			// special case for bess root itself
+			if node.ID == bessID {
+				continue
+			}
+
+			for _, module := range foundNodes {
+				if node.ID == graph.GenID(module) {
+					modFound = true
+					break
+				}
+			}
+			if !modFound {
+				log.Infof("Deleting node: %+v", *node)
+				p.Ctx.Graph.DelNode(node)
+			}
+		}
+
+		// now check edges
+		edges := p.Ctx.Graph.GetEdges(filter)
+		for _, edge := range edges {
+			edgeFound := false
+			for _, knownEdge := range foundEdges {
+				if edge == knownEdge {
+					edgeFound = true
+					break
+				}
+			}
+			if !edgeFound {
+				log.Infof("Deleting edge: %+v", *edge)
+				p.Ctx.Graph.DelEdge(edge)
+			}
+		}
+
+		p.Ctx.Graph.Unlock()
+		time.Sleep(1 * time.Second)
```

Comment:
any way to get an event instead of looping with a sleep ?

Reply:
I don't think there is a way to do this, but I will double check